source/link/linker.cpp

```diff
@@ -0,0 +1,755 @@
+                                        SPV_ERROR_INVALID_BINARY)
+             << "Failed to build a module out of " << modules.size()
+             << ".";
+    modules.push_back(std::move(module));
+  }
+
+  PassManager manager;
+
+  // Phase 1: Shift the IDs used in each binary so that they occupy a disjoint
+  //          range from the other binaries, and compute the new ID bound.
+  uint32_t id_bound = modules[0]->IdBound() - 1u;
+  for (auto i = modules.begin() + 1; i != modules.end(); ++i) {
+    Module* module = i->get();
+    module->ForEachInst([&id_bound](Instruction* insn) {
+      insn->ForEachId([&id_bound](uint32_t* id) { *id += id_bound; });
+    });
+    id_bound += module->IdBound() - 1u;
+    if (id_bound > 0x3FFFFF)
+      return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                        SPV_ERROR_INVALID_ID)
+             << "The limit of IDs, 4194303, was exceeded:"
+             << " " << id_bound << " is the current ID bound.";
+  }
+  ++id_bound;
+  if (id_bound > 0x3FFFFF)
+    return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                      SPV_ERROR_INVALID_ID)
+           << "The limit of IDs, 4194303, was exceeded:"
+           << " " << id_bound << " is the current ID bound.";
+
+  // Extract linking information
+  struct LinkingData {
+    size_t moduleIndex;
+    SpvId id;
+    SpvId typeId;
+    const char* name;
+    std::vector<SpvId> parametersIds;
+  };
+  std::vector<std::vector<LinkingData>> modulesImports(modules.size());
+  std::vector<std::unordered_map<std::string, LinkingData>> modulesExports(modules.size());
+  size_t moduleIndex = 0u;
+  for (auto& i : modules) {
+    // Figure out the imports and exports
+    for (auto& j : i->annotations()) {
+      if (j.opcode() == SpvOpDecorate && j.GetSingleWordInOperand(1u) == SpvDecorationLinkageAttributes) {
+        uint32_t type = j.GetSingleWordInOperand(3u);
+        SpvId id = j.GetSingleWordInOperand(0u);
+        LinkingData data;
+        data.moduleIndex = moduleIndex;
+        data.name = reinterpret_cast<const char*>(j.GetInOperand(2u).words.data());
+        data.id = id;
+        data.typeId = 0u;
+        for (const auto&k : *i) {
+          if (k.result_id() != id)
+            continue;
+          k.ForEachParam([&data](const Instruction* inst) {
+            data.parametersIds.push_back(inst->result_id());
+          });
+        }
+        if (type == SpvLinkageTypeImport)
+          modulesImports[moduleIndex].push_back(data);
+        else if (type == SpvLinkageTypeExport)
+          modulesExports[moduleIndex].emplace(data.name, data);
+      }
+      // TODO(pierremoreau): Can an SpvOpDecorateId be used for
+      //                     LinkageAttributes decorations?
+    }
+    ++moduleIndex;
+  }
+
+  //Find the import/export pairs
+  std::vector<std::pair<LinkingData, LinkingData>> linkingsToDo;
```

Comment:
Would really prefer a struct with sensible member names.  Hard to remember that first is the import and second is the export.